src/log/zipkin.cpp

```diff
@@ -0,0 +1,210 @@
+   //   std::string name;     // logical operation, should have low cardinality
+   //   std::string parentId; // The parent span id, or absent if root span
+   //   std::string id        // a-f0-9]{16}
+   //   int64_t     timestamp // epoch microseconds of start of span
+   //   int64_t     duration  // microseconds of span
+
+   uint64_t trace_id;
+   if( span.parent_id != 0 ) {
+      trace_id = span.parent_id;
+   } else {
+      trace_id = span.id;
+   }
+
+   fc::mutable_variant_object mvo;
+   mvo( "id", fc::to_hex( reinterpret_cast<const char*>(&span.id), sizeof( span.id ) ) );
+   mvo( "traceId", fc::to_hex( reinterpret_cast<const char*>(&trace_id), sizeof( trace_id ) ) );
+   if( span.parent_id != 0 ) {
+      mvo( "parentId", fc::to_hex( reinterpret_cast<const char*>(&span.parent_id), sizeof( span.parent_id ) ) );
+   }
+   mvo( "name", std::move( span.name ) );
+   mvo( "timestamp", span.start.time_since_epoch().count() );
+   mvo( "duration", (span.stop - span.start).count() );
+   mvo( "localEndpoint", fc::variant_object( "serviceName", service_name ) );
+
+   mvo( "tags", std::move( span.tags ) );
+   span.id = 0; // stop destructor of span from calling log again
+
+   // /api/v2/spans takes an array of spans
+   fc::variants result;
+   result.emplace_back( std::move( mvo ) );
+
+   return result;
+}
+
+void zipkin::log( zipkin_span::span_data&& span ) {
+   if( my->consecutive_errors > my->max_consecutive_errors || my->stopped )
+      return;
+
+   boost::asio::post(my->work_strand, [this, span{std::move(span)}]() mutable {
+      my->log( std::move( span ) );
+   });
+}
+
+void zipkin::impl::log( zipkin_span::span_data&& span ) {
+   if( consecutive_errors > max_consecutive_errors )
+      return;
+
+   try {
+      auto deadline = fc::time_point::now() + fc::microseconds( timeout_us );
+      if( !endpoint ) {
+         endpoint = url( zipkin_url );
+         dlog( "connecting to zipkin: ${p}", ("p", *endpoint) );
+      }
+
+      http.post_sync( *endpoint, create_zipkin_variant( std::move( span ), service_name ), deadline );
+
+      consecutive_errors = 0;
+      return;
+   } catch( const fc::exception& e ) {
+      wlog( "unable to connect to zipkin: ${u}, error: ${e}", ("u", zipkin_url)("e", e.to_detail_string()) );
+   } catch( const std::exception& e ) {
+      wlog( "unable to connect to zipkin: ${u}, error: ${e}", ("u", zipkin_url)("e", e.what()) );
+   } catch( ... ) {
+      wlog( "unable to connect to zipkin: ${u}, error: unknown", ("u", zipkin_url) );
+   }
+   ++consecutive_errors;
+}
+
+uint64_t zipkin_span::to_id( const fc::sha256& id ) {
+   // avoid 0 since id of 0 is used as a flag
+   return id._hash[3] == 0 ? 1 : id._hash[3];
+}
+
+zipkin_span::~zipkin_span() {
+   if( data.id == 0 )
+      return;
+   try {
+      if( zipkin_config::is_enabled() ) {
+         data.stop = time_point::now();
+         zipkin_config::get_zipkin().log( std::move( data ) );
```

Comment:
If somehow a span is destroyed prior to `::init` this would `throw` from the destructor.   I'm out of good ideas for how to "solve" this well but perhaps the constructor could check that zipkin is initialized so that it throws from there instead of a destructor?

Reply:
too little ☕  this morning.  it won't throw out of this destructor as you catch everything and drop it below.  I still think it may be better to throw out of the constructor if zipkin is not initialized so that errors such as this are not silent.

Reply:
I'm not sure I follow. `zipkin_config::is_enabled()` checks if `init()` has been called. Also zipkin_span constructor which generates an id will throw if not enabled. The zipkin_span constructor that does not generate an id will not throw, but its destructor will check `zipkin_config::is_enabled()` before calling `log`.